include/ui/color-mode.hpp

```diff
@@ -0,0 +1,133 @@
+  using GetColorFunction = std::function<std::uint32_t(
+      Optional<OutputComponent *>, Options &, size_t)>;
+  using GetPixelFromBufferFunction = std::function<std::uint32_t(
+      const MemoryValue &, size_t, Options &, size_t, size_t)>;
+  using GetColorFromBufferFunction = std::function<std::uint32_t(
+      const MemoryValue &, size_t, Options &, size_t)>;
+  using UpdateMemoryFunction = std::function<void(Optional<OutputComponent *>,
+                                                  Options &,
+                                                  std::shared_ptr<QImage>,
+                                                  size_t,
+                                                  size_t)>;
+  using UpdateAllPixelsFunction = std::function<void(
+      Optional<OutputComponent *>, Options &, std::shared_ptr<QImage>)>;
+  using UpdateAllColorsFunction = std::function<void(
+      Optional<OutputComponent *>, Options &, std::shared_ptr<QImage>)>;
+
+  /*
+   * \brief returns the color of the pixel at (x,y)
+   */
+  GetPixelFunction getPixel;
+  /*
+   * \brief returns the color from the color Table at the index index
+   */
+  GetColorFunction getColor;
+  /*
+   * \brief returns the color of the pixel at (x,y) fetching the data from the
+   *        prefetched Buffer buffer
+   */
+  GetPixelFromBufferFunction getPixelFromBuffer;
+  /*
+   * \brief returns the color from the color Table at the index index fetching
+   *        the data from the prefetched Buffer buffer
+   */
+  GetColorFromBufferFunction getColorFromBuffer;
+  /*
+   * \brief updates the image wherever the given change in memory modified the
+   *        image
+   */
+  UpdateMemoryFunction updateMemory;
+  /*
+   * \brief updates all pixels of the image
+   */
+  UpdateAllPixelsFunction updateAllPixels;
+  /*
+   * \brief updates all colors of the image
+   */
+  UpdateAllColorsFunction updateAllColors;
+
+  /*
+   * \brief loads a pointer from memory
+   * \param memoryAccess The access to the memory
+   * \param address Address of the pointer
+   * \param indirect ponter is at address : pointer is address
+   * \param cellSize size of a cell in memory
+   * \param pointerSize size of a pointer in Bits
+   * \return the pointer
+   */
+  static size_t loadPointer(Optional<OutputComponent *> memoryAccess,
+                            size_t address,
+                            bool indirect,
+                            size_t cellSize,
+                            size_t pointerSize,
+                            Options &o);
+  /*
+   * \brief returns the memoryValue from memory at address
+   * \param memoryAccess The access to the memory
+   * \param address address of the value
+   * \param length length of the value in byte
+   * \param defaultLength length of the output if there is no memoryAccess
+   * \return the memoryValue from memory at address
+   */
+  static MemoryValue getMemoryValueAt(Optional<OutputComponent *> memoryAccess,
```

Comment:
Why do you need an optional pointer? If you already need to use pointers you can probably use `null` as the invalid state